otaclient/app/ota_client_stub.py

```diff
@@ -14,661 +14,728 @@
+                    # NOTE(20230517): aligns with the previous behavior that create
+                    #                 response with RECOVERABLE OTA error for unresponsive
+                    #                 ECU.
                     response.add_ecu(
                         wrapper.UpdateResponseEcu(
-                            ecu_id=ecu_id,
+                            ecu_id=_ecu_contact.ecu_id,
                             result=wrapper.FailureType.RECOVERABLE,
                         )
                     )
             tasks.clear()
 
-        # after all subecus ack the update request, update my ecu
-        my_ecu_tracking_task: Optional[Coroutine] = None
-        if entry := request.find_update_meta(self.my_ecu_id):
-            if not self._ota_client.live_ota_status.request_update():
-                logger.warning(
-                    f"ota_client should not take ota update under ota_status={self._ota_client.live_ota_status.get_ota_status()}"
-                )
-                response.add_ecu(
-                    wrapper.UpdateResponseEcu(
-                        ecu_id=self.my_ecu_id,
-                        result=wrapper.FailureType.RECOVERABLE,
-                    )
-                )
-            else:
-                logger.info(f"update my_ecu: {self.my_ecu_id=}, {entry=}")
-                # dispatch update to local otaclient
-                _loop = asyncio.get_running_loop()
-                _loop.run_in_executor(
-                    self._executor,
-                    partial(
-                        self._ota_client.update,
-                        entry.version,
-                        entry.url,
-                        entry.cookies,
-                        fsm=self._update_session.fsm,
-                    ),
-                )
-
-                # launch the local update tracker
-                # NOTE: pass otaclient's get_last_failure method into the tracker
-                my_ecu_tracking_task = _RequestHandlingSession.my_ecu_update_tracker(
-                    executor=self._executor,
-                    fsm=self._update_session.fsm,
-                )
-
-                response.add_ecu(
-                    wrapper.UpdateResponseEcu(
-                        ecu_id=self.my_ecu_id,
-                        result=wrapper.FailureType.NO_FAILURE,
-                    )
+        # second: dispatch update request to local if required by incoming request
+        if update_req_ecu := request.find_ecu(self.my_ecu_id):
+            _resp_ecu = wrapper.UpdateResponseEcu(ecu_id=self.my_ecu_id)
+            try:
+                await self._otaclient_stub.dispatch_update(update_req_ecu)
+                update_acked_ecus.add(self.my_ecu_id)
+            except OTAClientBusy as e:
+                logger.warning(f"self ECU is busy, ignore local update request: {e!r}")
+                _resp_ecu.result = wrapper.FailureType.RECOVERABLE
+            response.add_ecu(_resp_ecu)
```

Comment:
@Bodong-Yang 
Let me confirm when dispatch_update raises OTAClientBusy.
In that case, that ecu is in updating but the status become failure?
I need a state transition chart or something.

Reply:
@obi-t4 

> Let me confirm when dispatch_update raises OTAClientBusy.

When local ECU is already doing an OTA update. ECU can only take one OTA at a time, other OTA update requests during on-going update will be rejected.

> In that case, that ecu is in updating but the status become failure?

This part of code is preparing response for incoming OTA update request. If local ECU is already in UPDATING ota_status, but receives another OTA update request, **the response for this OTA update request** will be RECOVERABLE. No ota_status change will occur as the OTA update is rejected.

This behavior(response RECOVERABLE FAILURE in update request response when OTA is busy) is introduced since we introduce multiple ECU OTA(from v3.0.0) I think.

Reply:
Perfect! Thank you!